game-core/src/test/java/games/strategy/triplea/delegate/battle/MustFightBattleTest.java

```diff
@@ -74,4 +96,293 @@ void testFightWithBothZeroStrength() {
     assertEquals(2, celebes.getUnitCollection().size());
     thenGetRandomShouldHaveBeenCalled(bridge, times(0));
   }
+
+  @ParameterizedTest(name = "{index} {0}")
+  @MethodSource
+  void attackerRetreatTerritories(
+      final String displayName,
+      final GamePlayer attacker,
+      final Map<Territory, Collection<Unit>> attackingFromMap,
+      final List<Unit> attackingUnits,
+      final Territory battleSite,
+      final BattleTracker battleTracker,
+      final GameData gameData,
+      final boolean headless,
+      final Collection<Territory> expected) {
+
+    final MustFightBattle battle =
+        new MustFightBattle(battleSite, attacker, gameData, battleTracker);
```

Comment:
Instantiating a `MustFightBattle` is not an easy thing.  It's code isn't really set up for Unit testing.  Most of the mocks that exist are just to work on the code that is called in the constructor.

One idea is to create a mock and then use mockito's `thenCallRealMethod` to call it:
```
MustFightBattle battle = mock(MustFightBattle.class);
battle.attackingUnits = attackingUnits;
... more setup ...
when(battle.getAttackerRetreatTerritories()).thenCallRealMethod();
Collection<Territory> retreatTerritories = battle.getAttackerRetreatTerritories();
```

I haven't tried that but I think it would work.

Another idea is to move this method to a helper class (call it `RetreatTerritories` and put both offensive and defensive retreat logic there) and then call the help class from the original `getAttackerRetreatTerritories`.  Then the test can work on that class instead of `MustFightBattle`.